### ISSUE TITLE
Trusted vendors

### ISSUE THREAD

**Issue by JakubMatejka** [keboola/developer-portal#200]
- Vendor may be marked as trusted
- Requirements:
  - Give us access to private repositories
  - Integration tests for components
- Each new app version of non-trusted vendors has to be manually approved by us

**Comment by JakubMatejka** [comment]
- [ ] vendorům přibude flag `trusted`
- [ ]  nový verze aplikace musíme schválit než se zveřejní

**Comment by odinuv** [comment]
to schvalovani nove verze by se melo tykat jen public appek

**Comment by JakubMatejka** [comment]
Jasně a to automatický chvalování u trusted vendorů bude jen pro nový verze, nebo i celý nový komponenty?

**Comment by odinuv** [comment]
asi bych dal jen novy verze

**Comment by JakubMatejka** [comment]
@Halama  Jestli teda dobře chápu https://keboola.slack.com/archives/C02C3GZUW/p1527185674000341 ... 

Dneska je tam tabulka `apps`, která se mění vždy při updatu a ze které se čte při listování n. dotazu na detail appky. Kromě ní je tam tabulka `appVersions`, do které se při každé úpravě appky přidá novej řádek a inkrementuje se číslo verze. Z té se čte, jen když se dotazuje na výpis verzí appky. V `apps` je tak to samý jako v `appVersions` pro poslední verzi appky.

Pokud to mám upravit jak navrhuješ, tak vlastně ve výsledku můžu tabulku `apps` úplně zrušit. Úprava appky přidá do `appVersions` novou verzi jako teď, schválení publikace nastaví `isPublic=1` na nejvyšší verzi a `isPublic=0` všem ostatním verzím. Výpis všech publikovaných appek tak bude jednoduchý select kde `isPublic=1` a výpis posledních verzí appek bude left outer joinem na sebe (něco jako `SELECT a.* FROM appVersions a LEFT OUTER JOIN appVersions b ON a.id = b.id AND a.version < b.version WHERE b.id IS NULL;`)

**Comment by Halama** [comment]
Asi bych začal tím jak se to bude chovat z pohledu API.

Public [list apps](https://kebooladeveloperportal.docs.apiary.io/#reference/0/admin-apps/list-apps) by měl vracet schválené verze appek tam je to asi jasný a to samé pro public detail appk.y

Jak by to mělo být v adminu? [admin detail appky](https://kebooladeveloperportal.docs.apiary.io/#reference/0/admin-apps/get-app-detail) by asi měl vracet poslední verzi appky. Pro detail publikované verze by mohl být parametr?  Co když nebude mít žádnou publikovanou verzi (v tuhle chvíli to znamená že je "private")? Mělo by to vrátit v tomhle případě poslendí verzi?
 
Taky bude potřeba vymyslet jak na to bezpečně přejít v syncu KBC. V tomhle případě by ale asi stačil release dev portalu kdy by se zatím všechny udpaty schvalovaly automaticky jako doteď a pak nasazení úpravy syncu kbc které by tahalo publikované verze.

Přemýšlím jestli to nerozdělit na dva flagy. jeden který říká že appka je veřejná a druhý který ukazuje na poslední schválenou verzi. Fungovalo by to potom konzistentě a bylo by případně možné schvalovat i privátní appky (což zatím není potřeba).

Jinak pokud by se to API chovalo takhle tak myslím že to jak to popisuješ na to sedí.

**Comment by Halama** [comment]
btw. ještě přemýšlím jestli dokážeme non-trusted vendorům zabránit aby přepsali už publikovaný tag v ECR? Samozřejmě to jenom můžeme dát do "podmínek" že tohle se prostě nedělá.

**Comment by tomasfejfar** [comment]
No, to jsi vytahnul zajímavou věc. My bychom to mohli udělat tak, že bysme nedeployovali tag, ale hash. Resp. zapsali bysme si aktuální hash, kterému odpovídá v okamžiku publishnutí ten tag. A deployovali ten. Ten se změní jen v případě, že se změní obsah a tím pádem chceme nové review, že je to OK.

**Comment by Halama** [comment]
jo pokud bysme u verze té appky měli evidovaný hash tak to by to docela pěkně řešilo.

**Comment by Halama** [comment]
1. ok
2. ten sync bych klidně nechal jak je teď jenom by to volalo z tím parametrem at to vrací tu publikovanou? Spíš bych se pak zaměřil na to aby se KBC syncovalo automaticky, nějak by se subscrinulo ke změnám v dev portálu a podle toho by se updatovalo. Tech 5 min je často hodně otravných:) Tohle je ale na samostatné issue.
3. asi je to blbost, neřešmě to:)
4. Pokud si vygeneruju pro appku [docker credentials](https://kebooladeveloperportal.docs.apiary.io/#reference/0/app/get-credentials-to-ecr-repository) můžu tam pod stejný tag opakovaně pushovat co chci a nemusím měnit verzi v dev portalu.. To co navrhoval Tomáš by to myslím pěkně řešil. Ale vyžadovalo by to úpravy i v docker runneru atd. Založil bych kdyžtak taky jako nové issue.

**Comment by Halama** [comment]
Na tu 4. jsem založil issue #203

**Comment by JakubMatejka** [comment]
Á, jasně, už rozumím. Ten digest zní dobře. 👍

**Comment by Halama** [comment]
Na [offsite](https://www.wunderlist.com/#/tasks/4144007790) jsme si ujasnili:
- schvalování appek non-trusted vendorů - musí mít public zdrojáky
- trusted vendors - s nima bude podepsaná smlouva
- všichni stávající se switchnout na non-trusted
- připravované změny oznámit dostatečně dopředu

**Comment by odinuv** [comment]
zdrojaky nemusi byt public, ale pristupny nam, ne?

**Comment by Halama** [comment]
jo to asi jo, ale musíme mít jistotu že to co pushnul odpovídá tomu kódu což si teda nejsem dokážeme ověřit v obou případech.
